File: frontend/src/types/activity.ts
import {
  LogEntity_Action,
  LogEntity_Level,
} from "@/types/proto/v1/logging_service";
import { FieldId } from "../plugins";
import { t } from "../plugins/i18n";
import { ExternalApprovalEvent } from "./externalApproval";
import {
  DatabaseId,
  InstanceId,
  IssueId,
  PrincipalId,
  SheetId,
  StageId,
  TaskId,
} from "./id";
import { IssueStatus } from "./issue";
import { MemberStatus, RoleType } from "./member";
import { StageStatusUpdateType, TaskStatus } from "./pipeline";
import { ApprovalEvent } from "./review";
import { Advice } from "./sqlAdvice";
import { VCSPushEvent } from "./vcs";

export function activityName(action: LogEntity_Action): string {
  switch (action) {
    case LogEntity_Action.ACTION_ISSUE_CREATE:
      return t("activity.type.issue-create");
    case LogEntity_Action.ACTION_ISSUE_COMMENT_CREATE:
      return t("activity.type.comment-create");
    case LogEntity_Action.ACTION_ISSUE_FIELD_UPDATE:
      return t("activity.type.issue-field-update");
    case LogEntity_Action.ACTION_ISSUE_STATUS_UPDATE:
      return t("activity.type.issue-status-update");
    case LogEntity_Action.ACTION_PIPELINE_STAGE_STATUS_UPDATE:
      return t("activity.type.pipeline-stage-status-update");
    case LogEntity_Action.ACTION_PIPELINE_TASK_STATUS_UPDATE:
      return t("activity.type.pipeline-task-status-update");
    case LogEntity_Action.ACTION_PIPELINE_TASK_RUN_STATUS_UPDATE:
      return t("activity.type.pipeline-task-run-status-update");
    case LogEntity_Action.ACTION_PIPELINE_TASK_FILE_COMMIT:
      return t("activity.type.pipeline-task-file-commit");
    case LogEntity_Action.ACTION_PIPELINE_TASK_STATEMENT_UPDATE:
      return t("activity.type.pipeline-task-statement-update");
    case LogEntity_Action.ACTION_PIPELINE_TASK_EARLIEST_ALLOWED_TIME_UPDATE:
      return t("activity.type.pipeline-task-earliest-allowed-time-update");
    case LogEntity_Action.ACTION_MEMBER_CREATE:
      return t("activity.type.member-create");
    case LogEntity_Action.ACTION_MEMBER_ROLE_UPDATE:
      return t("activity.type.member-role-update");
    case LogEntity_Action.ACTION_MEMBER_ACTIVATE:
      return t("activity.type.member-activate");
    case LogEntity_Action.ACTION_MEMBER_DEACTIVE:
      return t("activity.type.member-deactivate");
    case LogEntity_Action.ACTION_PROJECT_REPOSITORY_PUSH:
      return t("activity.type.project-repository-push");
    case LogEntity_Action.ACTION_PROJECT_DATABASE_TRANSFER:
      return t("activity.type.project-database-transfer");
    case LogEntity_Action.ACTION_PROJECT_MEMBER_CREATE:
      return t("activity.type.project-member-create");
    case LogEntity_Action.ACTION_PROJECT_MEMBER_DELETE:
      return t("activity.type.project-member-delete");
    case LogEntity_Action.ACTION_PROJECT_DATABASE_RECOVERY_PITR_DONE:
      return t("activity.type.database-recovery-pitr-done");
  }
  console.assert(false, `undefined text for activity type "${action}"`);
  return "";
}

export type ActivityLevel = "INFO" | "WARN" | "ERROR";

export type ActivityIssueCreatePayload = {
  issueName: string;
};

// TaskRollbackBy records an issue rollback activity.
// The task with taskID in IssueID is rollbacked by the task with RollbackByTaskID in RollbackByIssueID.
export type TaskRollbackBy = {
  issueId: IssueId;
  taskId: TaskId;
  rollbackByIssueId: IssueId;
  rollbackByTaskId: TaskId;
};

export type ActivityIssueCommentCreatePayload = {
  externalApprovalEvent?: ExternalApprovalEvent;
  issueName: string;
  taskRollbackBy?: TaskRollbackBy;
  approvalEvent?: ApprovalEvent;
};

export type ActivityIssueFieldUpdatePayload = {
  fieldId: FieldId;
  oldValue?: string;
  newValue?: string;
  issueName: string;
};

export type ActivityIssueStatusUpdatePayload = {
  oldStatus: IssueStatus;
  newStatus: IssueStatus;
  issueName: string;
};
export type ActivityStageStatusUpdatePayload = {
  stageId: StageId;
  stageStatusUpdateType: StageStatusUpdateType;
  issueName: string;
  stageName: string;
};

export type ActivityTaskStatusUpdatePayload = {
  taskId: TaskId;
  oldStatus: TaskStatus;
  newStatus: TaskStatus;
  issueName: string;
  taskName: string;
};

export type ActivityTaskFileCommitPayload = {
  taskId: TaskId;
  vcsInstanceUrl: string;
  repositoryFullPath: string;
  branch: string;
  filePath: string;
  commitId: string;
};

export type ActivityTaskStatementUpdatePayload = {
  taskId: TaskId;
  oldStatement: string;
  newStatement: string;
  oldSheetId: SheetId;
  newSheetId: SheetId;
  issueName: string;
  taskName: string;
};

export type ActivityTaskEarliestAllowedTimeUpdatePayload = {
  taskId: TaskId;
  oldEarliestAllowedTs: number;
  newEarliestAllowedTs: number;
  issueName: string;
  taskName: string;
};

export type ActivityMemberCreatePayload = {
  principalId: PrincipalId;
  principalName: string;
  principalEmail: string;
  memberStatus: MemberStatus;
  role: RoleType;
};

export type ActivityMemberRoleUpdatePayload = {
  principalId: PrincipalId;
  principalName: string;
  principalEmail: string;
  oldRole: RoleType;
  newRole: RoleType;
};

export type ActivityMemberActivateDeactivatePayload = {
  principalId: PrincipalId;
  principalName: string;
  principalEmail: string;
  role: RoleType;
};

export type ActivityProjectRepositoryPushPayload = {
  pushEvent: VCSPushEvent;
  issueId?: number;
  issueName?: string;
};

export type ActivityProjectDatabaseTransferPayload = {
  databaseId: number;
  databaseName: string;
};

export type ActivitySQLEditorQueryPayload = {
  statement: string;
  durationNs: number;
  instanceId: InstanceId;
  instanceName: string;
  databaseId: DatabaseId;
  databaseName: string;
  error: string;
  adviceList: Advice[];
};

export type ActionPayloadType =
  | ActivityIssueCreatePayload
  | ActivityIssueCommentCreatePayload
  | ActivityIssueFieldUpdatePayload
  | ActivityIssueStatusUpdatePayload
  | ActivityTaskStatusUpdatePayload
  | ActivityTaskFileCommitPayload
  | ActivityTaskStatementUpdatePayload
  | ActivityTaskEarliestAllowedTimeUpdatePayload
  | ActivityMemberCreatePayload
  | ActivityMemberRoleUpdatePayload
  | ActivityMemberActivateDeactivatePayload
  | ActivityProjectRepositoryPushPayload
  | ActivityProjectDatabaseTransferPayload
  | ActivitySQLEditorQueryPayload;

export interface FindActivityMessage {
  resource?: string;
  creatorEmail?: string;
  level?: LogEntity_Level[];
  action?: LogEntity_Action[];
  createdTsAfter?: number;
  createdTsBefore?: number;
  order?: "asc" | "desc";
  pageSize?: number;
  pageToken?: string;
}

File: backend/legacyapi/activity.go
package api

import (
	"github.com/bytebase/bytebase/backend/plugin/advisor"
	"github.com/bytebase/bytebase/backend/plugin/vcs"
)

// ActivityType is the type for an activity.
type ActivityType string

const (
	// Issue related.

	// ActivityIssueCreate is the type for creating issues.
	ActivityIssueCreate ActivityType = "bb.issue.create"
	// ActivityIssueCommentCreate is the type for creating issue comments.
	ActivityIssueCommentCreate ActivityType = "bb.issue.comment.create"
	// ActivityIssueFieldUpdate is the type for updating issue fields.
	ActivityIssueFieldUpdate ActivityType = "bb.issue.field.update"
	// ActivityIssueStatusUpdate is the type for updating issue status.
	ActivityIssueStatusUpdate ActivityType = "bb.issue.status.update"
	// ActivityIssueApprovalNotify is the type for notifying issue approval.
	ActivityIssueApprovalNotify ActivityType = "bb.issue.approval.notify"
	// ActivityPipelineStageStatusUpdate is the type for stage begins or ends.
	ActivityPipelineStageStatusUpdate ActivityType = "bb.pipeline.stage.status.update"
	// ActivityPipelineTaskStatusUpdate is the type for updating pipeline task status.
	ActivityPipelineTaskStatusUpdate ActivityType = "bb.pipeline.task.status.update"
	// ActivityPipelineTaskFileCommit is the type for committing pipeline task file.
	ActivityPipelineTaskFileCommit ActivityType = "bb.pipeline.task.file.commit"
	// ActivityPipelineTaskStatementUpdate is the type for updating pipeline task SQL statement.
	ActivityPipelineTaskStatementUpdate ActivityType = "bb.pipeline.task.statement.update"
	// ActivityPipelineTaskEarliestAllowedTimeUpdate is the type for updating pipeline task the earliest allowed time.
	ActivityPipelineTaskEarliestAllowedTimeUpdate ActivityType = "bb.pipeline.task.general.earliest-allowed-time.update"

	// Member related.

	// ActivityMemberCreate is the type for creating members.
	ActivityMemberCreate ActivityType = "bb.member.create"
	// ActivityMemberRoleUpdate is the type for updating member roles.
	ActivityMemberRoleUpdate ActivityType = "bb.member.role.update"
	// ActivityMemberActivate is the type for activating members.
	ActivityMemberActivate ActivityType = "bb.member.activate"
	// ActivityMemberDeactivate is the type for deactivating members.
	ActivityMemberDeactivate ActivityType = "bb.member.deactivate"

	// Project related.

	// ActivityProjectRepositoryPush is the type for pushing repositories.
	ActivityProjectRepositoryPush ActivityType = "bb.project.repository.push"
	// ActivityProjectDatabaseTransfer is the type for transferring databases.
	ActivityProjectDatabaseTransfer ActivityType = "bb.project.database.transfer"
	// ActivityProjectMemberCreate is the type for creating project members.
	ActivityProjectMemberCreate ActivityType = "bb.project.member.create"
	// ActivityProjectMemberDelete is the type for deleting project members.
	ActivityProjectMemberDelete ActivityType = "bb.project.member.delete"

	// SQL Editor related.

	// ActivitySQLEditorQuery is the type for executing query.
	ActivitySQLEditorQuery ActivityType = "bb.sql-editor.query"

	// SQL related.

	// ActivitySQLExport is the type for exporting SQL.
	ActivitySQLExport ActivityType = "bb.sql.export"

	// Database related.

	// ActivityDatabaseRecoveryPITRDone is the type for performing PITR on the database successfully.
	ActivityDatabaseRecoveryPITRDone ActivityType = "bb.database.recovery.pitr.done"
)

// ActivityLevel is the level of activities.
type ActivityLevel string

const (
	// ActivityInfo is the INFO level of activities.
	ActivityInfo ActivityLevel = "INFO"
	// ActivityWarn is the WARN level of activities.
	ActivityWarn ActivityLevel = "WARN"
	// ActivityError is the ERROR level of activities.
	ActivityError ActivityLevel = "ERROR"
)

// ActivityIssueCreatePayload is the API message payloads for creating issues.
// These payload types are only used when marshalling to the json format for saving into the database.
// So we annotate with json tag using camelCase naming which is consistent with normal
// json naming convention. More importantly, frontend code can simply use JSON.parse to
// convert to the expected struct there.
type ActivityIssueCreatePayload struct {
	// Used by inbox to display info without paying the join cost
	IssueName string `json:"issueName"`
}

// TaskRollbackBy records an issue rollback activity.
// The task with taskID in IssueID is rollbacked by the task with RollbackByTaskID in RollbackByIssueID.
type TaskRollbackBy struct {
	IssueID           int `json:"issueId"`
	TaskID            int `json:"taskId"`
	RollbackByIssueID int `json:"rollbackByIssueId"`
	RollbackByTaskID  int `json:"rollbackByTaskId"`
}

// ApprovalEvent duplicates store/approval.proto.
type ApprovalEvent struct {
	Status string `json:"status"`
}

// ActivityIssueCommentCreatePayload is the API message payloads for creating issue comments.
type ActivityIssueCommentCreatePayload struct {
	ExternalApprovalEvent *ExternalApprovalEvent `json:"externalApprovalEvent,omitempty"`

	TaskRollbackBy *TaskRollbackBy `json:"taskRollbackBy,omitempty"`

	ApprovalEvent *ApprovalEvent `json:"approvalEvent,omitempty"`

	// Used by inbox to display info without paying the join cost
	IssueName string `json:"issueName"`
}

// ActivityIssueFieldUpdatePayload is the API message payloads for updating issue fields.
type ActivityIssueFieldUpdatePayload struct {
	FieldID  IssueFieldID `json:"fieldId"`
	OldValue string       `json:"oldValue,omitempty"`
	NewValue string       `json:"newValue,omitempty"`
	// Used by inbox to display info without paying the join cost
	IssueName string `json:"issueName"`
}

// ActivityIssueStatusUpdatePayload is the API message payloads for updating issue status.
type ActivityIssueStatusUpdatePayload struct {
	OldStatus IssueStatus `json:"oldStatus,omitempty"`
	NewStatus IssueStatus `json:"newStatus,omitempty"`
	// Used by inbox to display info without paying the join cost
	IssueName string `json:"issueName"`
}

// ActivityIssueApprovalNotifyPayload is the API message payloads for notifying issue approval.
type ActivityIssueApprovalNotifyPayload struct {
	// ProtoPayload is protojson.Marshal(storepb.ActivityIssueApprovalNotifyPayload).
	ProtoPayload string `json:"protoPayload,omitempty"`
}

// ActivityPipelineStageStatusUpdatePayload is the API message payloads for stage status updates.
type ActivityPipelineStageStatusUpdatePayload struct {
	StageID               int                   `json:"stageId"`
	StageStatusUpdateType StageStatusUpdateType `json:"stageStatusUpdateType"`

	// Used by inbox to display info without paying the join cost
	IssueName string `json:"issueName"`
	StageName string `json:"stageName"`
}

// ActivityPipelineTaskStatusUpdatePayload is the API message payloads for updating pipeline task status.
type ActivityPipelineTaskStatusUpdatePayload struct {
	TaskID    int        `json:"taskId"`
	OldStatus TaskStatus `json:"oldStatus,omitempty"`
	NewStatus TaskStatus `json:"newStatus,omitempty"`
	// Used by inbox to display info without paying the join cost
	IssueName string `json:"issueName"`
	TaskName  string `json:"taskName"`
}

// ActivityPipelineTaskFileCommitPayload is the API message payloads for committing pipeline task files.
type ActivityPipelineTaskFileCommitPayload struct {
	TaskID             int    `json:"taskId"`
	VCSInstanceURL     string `json:"vcsInstanceUrl,omitempty"`
	RepositoryFullPath string `json:"repositoryFullPath,omitempty"`
	Branch             string `json:"branch,omitempty"`
	FilePath           string `json:"filePath,omitempty"`
	CommitID           string `json:"commitId,omitempty"`
}

// ActivityPipelineTaskStatementUpdatePayload is the API message payloads for pipeline task statement updates.
type ActivityPipelineTaskStatementUpdatePayload struct {
	TaskID       int    `json:"taskId"`
	OldStatement string `json:"oldStatement,omitempty"`
	NewStatement string `json:"newStatement,omitempty"`
	OldSheetID   int    `json:"oldSheetId,omitempty"`
	NewSheetID   int    `json:"newSheetId,omitempty"`
	// Used by inbox to display info without paying the join cost
	IssueName string `json:"issueName"`
	TaskName  string `json:"taskName"`
}

// ActivityPipelineTaskEarliestAllowedTimeUpdatePayload is the API message payloads for pipeline task the earliest allowed time updates.
type ActivityPipelineTaskEarliestAllowedTimeUpdatePayload struct {
	TaskID               int   `json:"taskId"`
	OldEarliestAllowedTs int64 `json:"oldEarliestAllowedTs,omitempty"`
	NewEarliestAllowedTs int64 `json:"newEarliestAllowedTs,omitempty"`
	// Used by inbox to display info without paying the join cost
	IssueName string `json:"issueName"`
	TaskName  string `json:"taskName"`
}

// ActivityMemberCreatePayload is the API message payloads for creating members.
type ActivityMemberCreatePayload struct {
	PrincipalID    int          `json:"principalId"`
	PrincipalName  string       `json:"principalName"`
	PrincipalEmail string       `json:"principalEmail"`
	MemberStatus   MemberStatus `json:"memberStatus"`
	Role           Role         `json:"role"`
}

// ActivityMemberRoleUpdatePayload is the API message payloads for updating member roles.
type ActivityMemberRoleUpdatePayload struct {
	PrincipalID    int    `json:"principalId"`
	PrincipalName  string `json:"principalName"`
	PrincipalEmail string `json:"principalEmail"`
	OldRole        Role   `json:"oldRole"`
	NewRole        Role   `json:"newRole"`
}

// ActivityMemberActivateDeactivatePayload is the API message payloads for activating or deactivating members.
type ActivityMemberActivateDeactivatePayload struct {
	PrincipalID    int    `json:"principalId"`
	PrincipalName  string `json:"principalName"`
	PrincipalEmail string `json:"principalEmail"`
	Role           Role   `json:"role"`
}

// ActivityProjectRepositoryPushPayload is the API message payloads for pushing repositories.
type ActivityProjectRepositoryPushPayload struct {
	VCSPushEvent vcs.PushEvent `json:"pushEvent"`
	// Used by activity table to display info without paying the join cost
	// IssueID/IssueName only exist if the push event leads to the issue creation.
	IssueID   int    `json:"issueId,omitempty"`
	IssueName string `json:"issueName,omitempty"`
}

// ActivityProjectDatabaseTransferPayload is the API message payloads for transferring databases.
type ActivityProjectDatabaseTransferPayload struct {
	DatabaseID int `json:"databaseId,omitempty"`
	// Used by activity table to display info without paying the join cost
	DatabaseName string `json:"databaseName,omitempty"`
}

// ActivitySQLEditorQueryPayload is the API message payloads for the executed query info.
type ActivitySQLEditorQueryPayload struct {
	// Used by activity table to display info without paying the join cost
	Statement  string `json:"statement"`
	DurationNs int64  `json:"durationNs"`
	InstanceID int    `json:"instanceId"`
	// DeprecatedInstanceName is deprecated and should be removed from future version.
	DeprecatedInstanceName string           `json:"instanceName"`
	DatabaseID             int              `json:"databaseId"`
	DatabaseName           string           `json:"databaseName"`
	Error                  string           `json:"error"`
	AdviceList             []advisor.Advice `json:"adviceList"`
}

// ActivitySQLExportPayload is the API message payloads for the exported SQL info.
type ActivitySQLExportPayload struct {
	// Used by activity table to display info without paying the join cost
	Statement    string `json:"statement"`
	DurationNs   int64  `json:"durationNs"`
	InstanceID   int    `json:"instanceId"`
	DatabaseID   int    `json:"databaseId"`
	DatabaseName string `json:"databaseName"`
	Error        string `json:"error"`
}
